model/src/main/kotlin/utils/PostgresFileArchiverStorage.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (C) 2021 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.model.utils
+
+import java.io.File
+
+import javax.sql.DataSource
+
+import kotlin.io.path.createTempFile
+
+import org.jetbrains.exposed.dao.id.IntIdTable
+import org.jetbrains.exposed.sql.Column
+import org.jetbrains.exposed.sql.Database
+import org.jetbrains.exposed.sql.ResultRow
+import org.jetbrains.exposed.sql.SchemaUtils.createMissingTablesAndColumns
+import org.jetbrains.exposed.sql.SchemaUtils.withDataBaseLock
+import org.jetbrains.exposed.sql.and
+import org.jetbrains.exposed.sql.insert
+import org.jetbrains.exposed.sql.select
+import org.jetbrains.exposed.sql.transactions.transaction
+
+import org.ossreviewtoolkit.model.Provenance
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.checkDatabaseEncoding
+import org.ossreviewtoolkit.model.utils.DatabaseUtils.tableExists
+import org.ossreviewtoolkit.utils.ORT_NAME
+
+/**
+ * A PostgreSQL based storage for archive files.
+ */
+class PostgresFileArchiverStorage(
+    /**
+     * The JDBC data source to obtain database connections.
+     */
+    dataSource: DataSource
+) : FileArchiverStorage {
+    init {
+        Database.connect(dataSource).apply { defaultFetchSize(1000) }
+
+        transaction {
+            withDataBaseLock {
+                if (!tableExists(FileArchiveTable.tableName)) {
+                    checkDatabaseEncoding()
+                    createMissingTablesAndColumns(FileArchiveTable)
+                }
+            }
+
+            commit()
+        }
+    }
+
+    override fun hasArchive(provenance: Provenance): Boolean {
+        checkIsUniqueStorageKey(provenance)
+
+        return transaction { queryFileArchive(provenance) } != null
+    }
+
+    override fun addArchive(provenance: Provenance, zipFile: File) {
+        checkIsUniqueStorageKey(provenance)
+
+        transaction {
+            if (hasArchive(provenance)) return@transaction
+
+            val fileArchive = FileArchive(provenance.storageKey(), zipFile.readBytes())
+            insertFileArchive(fileArchive)
+        }
+    }
+
+    override fun getArchive(provenance: Provenance): File? {
+        checkIsUniqueStorageKey(provenance)
+
+        val fileArchive = transaction { queryFileArchive(provenance) } ?: return null
+
+        val file = createTempFile(ORT_NAME, ".zip").toFile()
```

Comment:
This could leave a dangling file if the `writeBytes()` call throws an exception.